core/state/database.go

```diff
@@ -255,23 +256,106 @@ func (tds *TrieDbState) Copy() *TrieDbState {
 
 	if debug.IsIntermediateTrieHash() {
 		cpy.tp.SetUnloadNodeFunc(cpy.putIntermediateHash)
-		cpy.t.SetOnDeleteSubtreeFunc(tds.markSubtreeEmptyInIntermediateHash)
 		cpy.tp.SetCreateNodeFunc(cpy.delIntermediateHash)
 	}
 
 	return &cpy
 }
 
-func (tds *TrieDbState) markSubtreeEmptyInIntermediateHash(prefix []byte) {
-	if len(prefix) != common.HashLength {
+func ClearTombstonesForReCreatedAccount(db ethdb.MinDatabase, addrHash common.Hash) error {
+	addrHashBytes := addrHash[:]
+	if ok, err := HasTombstone(db, addrHashBytes); err != nil {
+		return err
+	} else if !ok {
+		return nil
+	}
+
+	buf := pool.GetBuffer(common.HashLength * 2)
+	defer pool.PutBuffer(buf)
+	copy(buf.B, addrHashBytes)
+
+	i := common.HashLength
+	for j := 0; j < 256; j++ {
+		buf.B[i] = uint8(j)
+		if err := db.Put(dbutils.IntermediateTrieHashBucket, buf.B[:i+1], []byte{}); err != nil {
+			return err
+		}
+	}
+
+	if err := db.Delete(dbutils.IntermediateTrieHashBucket, addrHashBytes); err != nil {
+		return err
+	}
+	return nil
+}
+
+// PutTombstoneForDeletedAccount - placing tombstone only if given account has storage in database
+func PutTombstoneForDeletedAccount(db ethdb.MinDatabase, addrHash []byte) {
+	if len(addrHash) != common.HashLength {
 		return
 	}
 
-	if err := tds.db.Put(dbutils.IntermediateTrieHashBucket, prefix, []byte{}); err != nil {
+	v, _ := db.Get(dbutils.AccountsBucket, addrHash)
+	if v == nil {
+		return
+	}
+	acc := accounts.NewAccount()
+	if err := acc.DecodeForStorage(v); err != nil {
```

Comment:
This decoding is only to check for empty root. There might be a more efficient way to do it (just check the first byte where the flag is)